File: performance/ack/ack.js
/** JavaScipt Ackermann Function */
const ack = (m, n) => {
    if (m == 0) return n + 1;
    if ((m > 0) && (n == 0)) return ack(m - 1, 1);
    if ((m > 0) && (n > 0)) return ack(m - 1, ack(m, n - 1));
}

/** JavaScript Benchaming method */
const bench = iters => {
    let min = Infinity;
    let max = 0;
    let sum = 0n;

    for (let i = 0; i < iters; i++) {
        const t_start = process.hrtime.bigint();
        ack(3, 6);
        const t_duration = (process.hrtime.bigint() - t_start) / 1000n;

        sum += t_duration;
        if (t_duration < min) min = t_duration;
        if (t_duration > max) max = t_duration;
    }

    console.log(`Average: ${sum / BigInt(iters)}us`);
    console.log(`Min: ${min}us`);
    console.log(`Max: ${max}us`);
}

console.log('\n=> JavaScript');
bench(100);
console.log();

File: performance/ack/ack.py
import time

# Python Implementation of Ackermann Function
def ack(m, n):
    if m == 0:
        return n + 1
    if (m > 0) and (n == 0):
        return ack(m - 1, 1)
    if (m > 0) and (n > 0):
        return ack(m - 1, ack(m, n - 1))
    pass


# Python Benchmarker
def bench(iters):
    min = float("inf")
    max = 0
    sum = 0

    for i in range(0, iters):
        start = time.time()
        ack(3, 6)  # NOTE: Python cannot exceed ack(3, 7)
        elapsed = time.time() - start  # this is in seconds

        sum = sum + elapsed
        if elapsed < min:
            min = elapsed
        elif elapsed > max:
            max = elapsed

    print("Average: " + str((sum / iters) * 1000) + "ms")
    print("Min: " + str(min * 1000) + "ms")
    print("Max: " + str(max * 1000) + "ms")
    pass


print("\n=> Python3")
bench(100)
print()
